src/services/completions.ts

```diff
@@ -2905,7 +2905,10 @@ function getCompletionData(
     log("getCompletionData: Semantic work: " + (timestamp() - semanticStart));
     const contextualType = previousToken && getContextualType(previousToken, position, sourceFile, typeChecker);
 
-    const literals = mapDefined(
+    // don't include literal suggestions after <input type="text" [||] /> (#51667) and after quote (#52675)
+    // for strings getStringLiteralCompletions handles completions
+    const isLiteralExpected = !isStringLiteralLike(previousToken) && !isJsxIdentifierExpected;
+    const literals = !isLiteralExpected ? [] : mapDefined(

```

Comment:
Basically it always disallow literal completions in string and `identifier` locations. I wonder wether it can any edge cases I don't know about?

Reply:
You could instead use `isInStringOrRegularExpressionOrTemplateLiteral` which will tell you if you're *inside* of the node (even if that node is unterminated).

https://github.com/microsoft/TypeScript/blob/1c822c42a4d527c2f97a27cf1d5cfe2e5558e280/src/services/completions.ts#L3694

Reply:
@DanielRosenwasser so as I understand your proposal is to replace `isStringLiteralLike` with `isInStringOrRegularExpressionOrTemplateLiteral`, but whats the difference? Am I missing something?

Reply:
If I understand correctly, the second one checks if the position is inside of the string, whereas the original just checks if the previous token is a string or template string. You might still want to keep `isStringLiteralLike`, but you want to check if the cursor is actually _inside of_ the string.

Reply:
>  but you want to check if the cursor is actually inside of the string

But do I need this check? What issue it covers? I'm pretty sure there is already check if I'm inside of the string https://github.com/microsoft/TypeScript/blob/d0938c85ec65816448d730e9342f65620dbb857a/src/services/stringCompletions.ts#L198-L200

> If I understand correctly, the second one checks if the position is inside of the string, whereas the original just checks if the previous token is a string or template string

I have two checks for two issues correspondly, first one checks if the previous token is a string and second checks if jsx attribute identifier expected (see added tests). In the second check token is not necessarily is string: [playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3gMSUgXPABQDeAvoSfFIQAzwA+8AjADTwBG8ZAlPALwA+aigCeAWABQUgNwAeXEmr8StMvAD0AKi0bNgoA).

Reply:
I think I see - it sounds like the idea that the previous token the the one immediately to the left of the original context token, so we could not have been *inside* of the string. Is that right?

Reply:
`previousToken` is either the token you’re currently typing or one behind, if you closed out a token but haven’t started typing a new one yet:

```
"hello|  // previousToken = String Literal
"hello"| // previousToken = String Literal
```

but, when the cursor is inside the string, a different code path would have already handled that and returned results from `stringCompletions.ts`. So in this case, we know we’re not inside the string. (Confusingly, `stringCompletions.ts` doesn’t handle completions of entire string literals including the quotes; it only handles completions inside existing string literals.)

Reply:
Wow this is really great explanation of first check. Thank you so much, I wish I could write same explanations in first place 😄